### PR TITLE
[Core] Add saving distance in skin in `CalculateNodalDistanceToSkinProcess`

### PR DESCRIPTION
**Pull request by loumalouomega**
**📝 Description**

This PR introduces several enhancements and adjustments in `CalculateNodalDistanceToSkinProcess` (the process for calculating nodal distances to the skin).  Key updates include:

- **Distance value storage in skin model**: Implemented functionality to store distance values directly in the skin model part. By default is deactivated, if activated will save the maximum distance in the found skin geometry and set the flag `VISITED`. This also deactivate shared memory parallelization.

- **Lambda function updates**: Updated lambda functions now support the improved mechanism for saving distance values in the skin model part, facilitating a more streamlined and efficient process.

- **Default parameters**: Added and refined default parameters for the process to enhance user experience by providing clearer guidelines and reducing the need for manual adjustments.

- **Improved printed information**: Adjustments to the information printout provide a better representation of process settings and variables, aiding in debugging and process optimization.

- **Documentation updates**: The `calculate_nodal_distance_to_skin_process.md` documentation has been extensively updated to include detailed explanations of the process enhancements. It covers the support for calculating distances to skin entities, introduces new lambda functions, and discusses the addition of flags for data type selection and decision-making regarding the storage of distances in the skin model part. The documentation also clarifies the functionalities of default parameters.

- **New test method**: Introduced a new test method, `test_ComputeDistanceToSkinWithSavedDistanceInSkin`, to the `test_calculate_nodal_distance_to_skin_process.py` file. This test verifies that specific conditions in the skin model part are correctly addressed during the process, ensuring the reliability of the new features.

**🆕 Changelog**

- Added local flags for enhanced control and clarity, facilitating better customization of the process ([Commit Link](https://github.com/KratosMultiphysics/Kratos/commit/6650b3c2226d740008f18f72d28eca13b2a6e8c2)).

- Implemented functionality for saving maximum distance values to skin entities within the skin model part, optimizing the process by eliminating the need for parallel algorithms in this context ([Commit Link](https://github.com/KratosMultiphysics/Kratos/commit/187c670fb249ffd72a2d593a24790bb83c0b2459)).

- Added a new test case, `test_ComputeDistanceToSkinWithSavedDistanceInSkin`, to ensure the process accurately handles conditions within the skin model part, thereby verifying the enhancements made ([Commit Link](https://github.com/KratosMultiphysics/Kratos/commit/3eac8743bc97054d8314f049c628569ad6a8afb0)).

### CODE REVIEW

**Review by rubenzorrilla — COMMENTED** [review]
Considering that we have a parameters-based constructor, which is the advantage of using local flags rather than a simple bool? In this regard, I note that local flags have always been a pain in the *ss.

**Comment by loumalouomega** [comment]
> Considering that we have a parameters-based constructor, which is the advantage of using local flags rather than a simple bool? In this regard, I note that local flags have always been a pain in the *ss.

- Process is already a  class derived from Flags, so it is already there not adding anything to memory.
- Faster.
- Less memory consumption.
- It is already done, we had  a discussion back in the time of `from_json_check_result_process` implementation and we agreed that if more than one bool it was better this approach than adding more member variables. See https://github.com/KratosMultiphysics/Kratos/blob/4e44cbc551cfd2b7a9a172eed28eaf8b678ace92/kratos/processes/from_json_check_result_process.h#L66
- Local flags are painful if you plan to use it outside the class, that is not the case.

**Comment by rubenzorrilla** [comment]
> > Considering that we have a parameters-based constructor, which is the advantage of using local flags rather than a simple bool? In this regard, I note that local flags have always been a pain in the *ss.
> 
>     * Process is already a  class derived from Flags, so it is already there not adding anything to memory.
> 

I think this is a legacy reason. Honestly, I don't think it is the best of the ideas to make everything derive from `Flags`. Having said this, I'm of the opinion of stop exploiting this capability.

>     * Faster.
> 
>     * Less memory consumption.
> 

Well, I don't think performance of having two `bool` member variables is relevant considering the sort of operations we're doing in here. Indeed, I'd consider it negligible.

>     * It is already done, we had  a discussion back in the time of `from_json_check_result_process` implementation and we agreed that if more than one bool it was better this approach than adding more member variables. See https://github.com/KratosMultiphysics/Kratos/blob/4e44cbc551cfd2b7a9a172eed28eaf8b678ace92/kratos/processes/from_json_check_result_process.h#L66
> 

If so, I think we should reconsider this again @KratosMultiphysics/technical-committee .

>     * Local flags are painful if you plan to use it outside the class, that is not the case.

You are not planning to do so but by adding them you open the possibility for others to do it. Hence, I'm of the opinion of simply precluding it.

**Comment by loumalouomega** [comment]
> > > Considering that we have a parameters-based constructor, which is the advantage of using local flags rather than a simple bool? In this regard, I note that local flags have always been a pain in the *ss.
> > 
> >     * Process is already a  class derived from Flags, so it is already there not adding anything to memory.
> > 
> 
> I think this is a legacy reason. Honestly, I don't think it is the best of the ideas to make everything derive from `Flags`. Having said this, I'm of the opinion of stop exploiting this capability.
> 
> >     * Faster.
> > 
> >     * Less memory consumption.
> > 
> 
> Well, I don't think performance of having two `bool` member variables is relevant considering the sort of operations we're doing in here. Indeed, I'd consider it negligible.
> 
> >     * It is already done, we had  a discussion back in the time of `from_json_check_result_process` implementation and we agreed that if more than one bool it was better this approach than adding more member variables. See https://github.com/KratosMultiphysics/Kratos/blob/4e44cbc551cfd2b7a9a172eed28eaf8b678ace92/kratos/processes/from_json_check_result_process.h#L66
> > 
> 
> If so, I think we should reconsider this again @KratosMultiphysics/technical-committee .
> 
> >     * Local flags are painful if you plan to use it outside the class, that is not the case.
> 
> You are not planning to do so but by adding them you open the possibility for others to do it. Hence, I'm of the opinion of simply precluding it.
> 

The choice to use local flags instead of simple boolean variables is not merely a legacy decision; it was actively considered advantageous during the implementation of the Json check process. Local flags offer several benefits. 

1. **Integration with existing structures:** Since the `Process` class already derives from `Flags`, utilizing local flags does not introduce additional memory overhead. This integration leverages the existing infrastructure, promoting code consistency and reducing the need for redundant data structures.

2. **Efficiency:** Although the performance difference between using two boolean variables and flags might seem negligible for this particular case, the affirmation was done for a more general scenario. 

3. **Code maintenance and evolution:** The decision to use local flags was made after careful consideration, including discussions around the `from_json_check_result_process` implementation. This approach was seen as preferable for managing multiple boolean conditions without cluttering the class with additional member variables. Revisiting this decision should be based on a comprehensive evaluation of current and future requirements, not just on perceived legacy or simplicity.

4. **Flexibility and use cases:** Local flags have been utilized in numerous implementations within KratosMultiphysics, demonstrating their utility beyond a single case. 

In fact, for the particular case ofpProcess flags, I use them in many of my implementations, and I am not the only one, and usually it is done with local flags. Local flags are not that problematic. 

5. **Risk management:** Concerns about misuse or overextension by using local flags can be addressed through documentation, code reviews, and establishing best practices rather than avoiding their use altogether. Similar to other powerful features in programming languages (e.g., `const_cast`), the focus should be on educating developers on proper use rather than limiting functionality based on potential misuse.

In conclusion, local flags represent a deliberate design choice that aligns with the architectural and operational requirements of Kratos. They offer a structured and efficient way to manage multiple condition flags within classes, enhancing code clarity and performance. The decision to utilize them should be guided by their proven benefits and applicability to the task at hand, rather than a blanket rejection based on concerns over potential misuse.

**Comment by pooyan-dadvand** [comment]
I tend to agree with @rubenzorrilla that making the constructor incompatible with `(model,parameter)` API in this case is not well justified. The process is very complex, needs lot of memory for bins and others and is very time consuming. So conversion of parameters to an internal bool is not significant.

**Comment by loumalouomega** [comment]
> Just to add that local flags are suitable as nodal or elemental data (where the memory and access time are crucial) but in processes it is more a legacy from the time we didn't have parameters.

OK, but there are used in different places

**Comment by loumalouomega** [comment]
> I tend to agree with @rubenzorrilla that making the constructor incompatible with `(model,parameter)` API in this case is not well justified. The process is very complex, needs lot of memory for bins and others and is very time consuming. So conversion of parameters to an internal bool is not significant.

I am using `(model,parameter)`, I think some information was mixed

**Comment by pooyan-dadvand** [comment]
Regarding to usage of local_flags in the processes, @KratosMultiphysics/technical-committee would encourage the use of Parameters in the argument list and not using Flags nor bools if possible. Internally, we see using member variables is more clear and straightforward.

**Comment by loumalouomega** [comment]
> Regarding to usage of local_flags in the processes, @KratosMultiphysics/technical-committee would encourage the use of Parameters in the argument list and not using Flags nor bools if possible. Internally, we see using member variables is more clear and straightforward.

The model part constructor is only accesible via C++, not python, do you want I removed it?, this was added less thana a week ago.

**Comment by loumalouomega** [comment]
> Regarding to usage of local_flags in the processes, @KratosMultiphysics/technical-committee would encourage the use of Parameters in the argument list and not using Flags nor bools if possible. Internally, we see using member variables is more clear and straightforward.

I have nveer used local flags in this PR as input argument, I was using them to store the information (Process derives from Flags)

File: docs/pages/Kratos/Processes/Calculation/calculate_nodal_distance_to_skin_process.md [hunk 1]
```diff
@@ -72,5 +85,8 @@ Defines the skin model part for distance computations. Like the volume model par
 ##### `distance_database`
 This flag indicating whether historical variable is enabled or otherwise non-historical is considered. It is set by string, options are `nodal_historical` or `nodal_non_historical`. Default value is `nodal_historical`.
 
+#### `save_distance_in_skin`
+This flag indicates if the distance is going to be saved in the skin. By default is false, if activated will save the maximum distance in the found skin geometry and set the flag `VISITED`. This also deactivate shared memory parallelization.
```
Comment (pooyan-dadvand) [inline docs/pages/Kratos/Processes/Calculation/calculate_nodal_distance_to_skin_process.md:89]: save_max_distance_in_skin would be more descriptive

File: kratos/processes/calculate_nodal_distance_to_skin_process.cpp [hunk 2]
```diff
@@ -92,7 +101,35 @@ void CalculateNodalDistanceToSkinProcess::Execute()
...
+            auto result = rBins.SearchNearest(rNode);
+            const double value = result.GetDistance();
+            auto p_result = result.Get().get();
+            if (p_result->GetValue(r_variable) < value) {
+                p_result->SetValue(r_variable, value);
+                p_result->Set(VISITED);
+            }
+            rNode.SetValue(r_variable, value);
+        }
+    };
```
Comment (pooyan-dadvand) [inline kratos/processes/calculate_nodal_distance_to_skin_process.cpp:121]: Wouldn't be cleaner and with less duplication if we only define a template method SetNodalValue and specialize it with historical and non historical template arguments

**Review by pooyan-dadvand — CHANGES_REQUESTED** [review]
Are you using the non standard constructor? If not, I would remove it

File: docs/pages/Kratos/Processes/Calculation/calculate_nodal_distance_to_skin_process.md (hunk 1 above)
Comment (loumalouomega) [reply docs/pages/Kratos/Processes/Calculation/calculate_nodal_distance_to_skin_process.md:89]: Done

File: kratos/processes/calculate_nodal_distance_to_skin_process.cpp (hunk 2 above)
Comment (loumalouomega) [reply kratos/processes/calculate_nodal_distance_to_skin_process.cpp:121]: Done

**Comment by loumalouomega** [comment]
Suggestions applied @pooyan-dadvand

**Comment by loumalouomega** [comment]
> Suggestions applied @pooyan-dadvand

This is ready @pooyan-dadvand

**Comment by loumalouomega** [comment]
> > Suggestions applied @pooyan-dadvand
> 
> This is ready @pooyan-dadvand

Ping @pooyan-dadvand

**Comment by loumalouomega** [comment]
> > > Suggestions applied @pooyan-dadvand
> > 
> > 
> > This is ready @pooyan-dadvand
> 
> Ping @pooyan-dadvand

👋

**Comment by loumalouomega** [comment]
ANyone?

**Comment by loumalouomega** [comment]
@pooyan-dadvand

**Comment by loumalouomega** [comment]
Ping @pooyan-dadvand

**Review by pooyan-dadvand — APPROVED** [review]
Thanks @loumalouomega
